lib/document.js

```diff
@@ -1072,18 +1072,19 @@ Document.prototype.$set = function $set(path, val, type, options) {
           this.$__.savedState[path + '.' + key] = priorVal[key];
         }
       }
+      if (!utils.isEmptyObject(val)) {
```

Comment:
This unfortunately breaks setting a nested field to an empty object (see test failures), so this isn't a correct solution. I haven't been able to look closely enough to have any suggestions about how to fix this issue though.

Reply:
@vkarpov15 thanks for the comment, 

I see the problem this fix creates.  
It is a blocker for us unfortunately without this fixed We can't update mongoose,  
it will be great if you can have a look and provide some help to fix this.

Reply:
@vkarpov15 

I had a deeper look at the repo, and I believe this time I find the correct solution for this.  

It was basically happening because of not restoring the _skipMinimizeTopLevel to false back after we set it to overwrite the nested object value if it is set by the user. 

```
 if (type === true &&
          !prefix &&
          path[key] != null &&
          pathtype === 'nested' &&
          this._doc[key] != null &&
          Object.keys(this._doc[key]).length === 0) {
        delete this._doc[key];
        // Make sure we set `{}` back even if we minimize re: gh-8565
        options = Object.assign({}, options, { _skipMinimizeTopLevel: true });
      } else {
        // Make sure we set `{_skipMinimizeTopLevel: false}` if don't have overwrite: gh-10441
        options = Object.assign({}, options, { _skipMinimizeTopLevel: false });
      }
```